README.md

            The AcCoRD Simulator
            (Actor-based Communication via Reaction-Diffusion)

This document is the README for AcCoRD v1.4.2 (2020-02-12)

# Introduction

Welcome to the AcCoRD Simulator (Actor-based Communication via Reaction-Diffusion). AcCoRD is a simulation tool for molecular communication. It is a hybrid microscopic/mesoscopic simulator that functions as a generic solver of stochastic advection-reaction-diffusion but which has been developed from the perspective of communications analysis. The focus is to efficiently generate the statistics of molecule observations at some location (i.e., at a receiver).

Quicklinks:
* User-friendly external webpage (includes links to feature list, download page, and instructions for use): https://warwick.ac.uk/fac/sci/eng/staff/ajgn/software/accord/
* Old AcCoRD releases: https://github.com/adamjgnoel/AcCoRD/releases
* Main AcCoRD development page on Github: https://github.com/adamjgnoel/accord
* Maintenance of existing issues and planned features: https://github.com/adamjgnoel/AcCoRD/issues
* Github wiki (just a copy of this README): https://github.com/adamjgnoel/AcCoRD/wiki

Primary features include:
* As a hybrid solver, AcCoRD integrates two simulation models to define 3D environments with flexible local accuracy.  Microscopic regions define each molecule individually and evolve over discrete time steps. Mesoscopic regions count the number of molecules in disjoint virtual bins (called subvolumes) and evolve over time steps with continuous granularity.
* "Actors" are either active molecule sources (i.e., transmitters) or passive observers (i.e., receivers). Active actors can modulate binary sequence data and release molecules over specified intervals. Passive actors record the number of molecules observed and can also record molecule positions.
* Chemical reactions can be defined locally or globally. A general framework for defining reactions can accommodate reactions such as molecule degradation, enzyme kinetics, reversible or irreversible surface binding, ligand-receptor binding, transitions across boundary membranes, and simplified molecular crowding. Surface binding reactions include absorption (consumption), adsorption (sticking), and desorption (release).
* Independent realizations of a simulation can be repeated an arbitrary number of times (and on different computers) and then aggregated to determine the average behavior and channel statistics.
* AcCoRD's interface has been designed to be helpful to the novice user while providing meaningful output. AcCoRD also includes post-processing tools developed in MATLAB. These tools enable saving aggregated simulation output files, plotting receiver observations (either the time-varying behavior or empirical distributions at specified times), and visualizing the physical environment (either as still images or compiled into a video)

## Audience

If you are a communications researcher who is interested in studying molecular communication, then you are the intended audience. However, anyone interested in the multi-scale simulation of reaction-diffusion systems might also find AcCoRD useful.

## Development Status

Development of AcCoRD is on-going and hosted on Github (https://github.com/adamjgnoel/AcCoRD).

Existing bugs and planned features are described on the Issues page: https://github.com/adamjgnoel/AcCoRD/issues

Most identified content on the Issues page are planned features.

# Publications

Publications that either describe or use AcCoRD can be found at this page: https://warwick.ac.uk/fac/sci/eng/staff/ajgn/software/accord/publications/

# Known Issues and Limitations

This list is current as of v1.3. See https://github.com/adamjgnoel/AcCoRD/issues for the latest details.
* Full 2D simulations are mostly untested
* 2nd order chemical reactions in microscopic regions need a binding radius specified.
* There is an option to set whether actor is independent but the value is ignored since dependent actors have not yet been implemented
* 2 regions must share an outer face in order to be classified as neighbors, unless one region is the parent of the other. So, if a child region is up against a face of its parent, and that parent also has a parent but doesn't share the same face with its parent, then the parent's parent will not be recognized as a neighbor of the child. In such a case molecule transitions between the child and the "grandparent" are not possible.
* Surface regions must be microscopic and can only have microscopic regions as neighbors.
* A child region is always assumed to be a neighbor of its parent, which is not strictly true but should not create a problem during a simulation
* microscopic molecules in a spherical region may go a very small distance beyond the region boundary but are still believe to be inside the sphere. This is due to numerical underflow, and might mean that the potential chemical reactions for the molecule are inconsistent with its location.
* A mesoscopic subvolume adjacent to a microscopic region needs that region to cover the entirety of every face that is adjacent to that region (i.e., part of the adjacent face should not also be adjacent to some other region). Otherwise the hybrid interface will not have proper transitions.
* Actors that act at the same time will do so in a random order (this is by design). So, if an active actor is adding molecules at the exact time that a passive actor is supposed to observe them, it is unknown a priori which will act first. It will depend on whichever is currently ranked higher in the timer heap. This could be modified in the future to rank the actors in the order they are defined in the event that their timer values are the same.
* Unbinding radii are only applied to second order microscopic reactions that have multiple products.


# Future Features

Development of AcCoRD is on-going. See https://github.com/adamjgnoel/AcCoRD/issues for the latest details. Here is a list of some planned end-user features (in no particular order):
* Expand flow model. Currently, flow must be steady and uniform (within a given region).
* Allow dependent actors, whose behavior depends on the observations of passive actors. This would allow implementation of more complex communication networks (e.g., relays)
* Add actors that track changes to molecule composition in their volumes
* Improve implementation of 2nd order chemical reactions in microscopic regions
* Enable surfaces to mesoscopic regions. Could be achieved via exclusion regions
* Write more MATLAB utility functions to perform common tasks with simulation output (e.g., bit error probabilities)
* Use different microscopic time steps in different microscopic regions
* Add checks on subvolume size relative to diffusion coefficients
* Add "tau leaping" in mesoscopic regime to improve scalability
* Add actor mobility, such that actor location can vary over time (either randomly or deterministically)
* Add more modulation schemes for active actors


# Installation

Download and installation instructions can be found on this page: https://warwick.ac.uk/fac/sci/eng/staff/ajgn/software/accord/downloads/


# Basic Usage

Usage instructions can be found in the User's Manual that is included with the software download


# Licensing

Main AcCoRD files are copyright 2016-2020 by Adam Noel under the "New BSD" license. For full details of licensing, including use of third-party code, please see LICENSE.txt

# Credits

Developed and maintained by Adam Noel (https://warwick.ac.uk/fac/sci/eng/staff/ajgn)

Testing: M. Halimeh (v0.2 to v0.4) and T. Schwering (v0.3 to v0.5)

Supervision and Support: K. C. Cheung (to v1.0), A. Hafid (v0.4 to v1.0), D. Makrakis (v0.4 to v1.1), and R. Schober  (to v1.0).

CHANGELOG.txt

            The AcCoRD Simulator
            (Actor-based Communication via Reaction-Diffusion)

This document is the CHANGELOG for AcCoRD v1.4.2 (2020-02-12)
- [NEW] Updated source code so that it can be compiled in MacOS systems. Executables are not provided but Mac OS users should be able to compile it themselves using the same syntax as that for Linux users
- [CHANGE] Removed build scripts and executables specific to RHEL/CentOS Linux. The same files can be used as with Debian/Ubuntu Linux, as long as the executables are compiled on a RHEL/CentOS Linux operating system. This change was made to streamline future updates.
- [FIX] Updated video generator code for compatibility from MATLAB R2019b. Should still be backwards compatible to R2016b.
- [FIX] Corrected indexing of plots where markers are chosen to draw the individual molecules in the simulated environment

Changes introduced in AcCoRD v1.4.1 (2019-04-16)
- [NEW] added a "Quick Plot" function for empty environments, accordEmptyEnvironmentQuick, which takes a single argument (path to configuration file) and plots all of the regions and actors with default plot settings. Corresponding note added to the user manual
- [FIX] Updated recommended MATLAB version to R2016b, because newer version of function to check strings is used when importing a configuration file (this call had changed some number of versions ago but compatibility with R2015a wasn't checked)

Changes introduced in AcCoRD v1.4 (2018-08-06)
- [NEW] added the A Priori Monte Carlo (APMC) Absorption algorithm, which enables accurate simulation with much larger time steps. Included as a new option for surface reactions. This initial implementation requires that the corresponding reaction rate must be infinite, otherwise the reaction will never occur. A threshold can constrain its use for additional runtime efficiency.
- [NEW] added a sample configuration file that uses the APMC absorption algorithm. Simulated environment has a point transmitter and a single absorbing sphere.
- [NEW] added a "Quick Plot" function, accordQuickPlot, which takes a single argument and plots all time-varying average observations by all passive actors. It is implemented as a simpler alternative to accordPlotMakerWrapper. It can also be used to import AcCoRD output, but in such a case it will only load data from the first corresponding seed file.
- [FIX] corrected check on molecule transitions between parent and child regions when a molecule is initialized directly at the parent/child boundary
- [FIX] fixed the loading of surface reaction parameters when there are products that are not released away from the surface.
- [FIX] corrected missing assignment for first order reactions that don't have any product molecules to place. Led to memory error when first molecule in a list has to be removed 

Changes introduced in AcCoRD v1.3 (2018-07-31)
- [NEW] added PDF User's Manual to replace the helper text files and many pages on the AcCoRD homepage, to make it easier to expand and update as needed.
- [ENHANCEMENT] modified default list of actors and regions to plot so that EVERYTHING is plotted and not just the first region and first actor. This default list is more user friendly.
- [ENHANCEMENT] added checks on some active actor parameters because they are only conditionally necessary
- [FIX] fixed detection of actors that are defined as the union of more than 1 region. JSON imported the list of regions as a string array and not a cell array
- [FIX] fixed bug that prevent an environment from being plotted if the plot list included a point region or actor. A new warning is added to state that point actors and regions are not plotted

Changes introduced in AcCoRD v1.2 (2018-05-30)
- [NEW] added plotting molecules as 3D shapes instead of only as markers, which is useful for visualizing depth. Added the sphere shape as an option.
- [NEW] online documentation migrated to University of Warwick
- [FIX] corrected implementation of replication reactions, where a first order reactant produces at least one copy of itself. If such a reactant had reacted again within the same microscopic time step, then the new molecule(s) previously went missing.
- [FIX] corrected calculation of diffusion propensity when more than 1 molecule is added to a subvolume
- [FIX] corrected limit on the number of actions made by passive actors.
- [FIX] updated code that writes to MATLAB output file so that filename is not cropped if "fileName" has a period in it
- [FIX] corrected the range generated for 3D Empirical PMFs (was off by 1).
- [FIX] fixed typo in comment describing implementation of reactions for "new" microscopic molecules (created in current time step).

Changes introduced in AcCoRD v1.1 (2016-12-24)
- [NEW] added net fluid flow. Can be defined globally and in individual regions. Each molecule can have its own flow specification. Currently, only type of flow available is steady uniform flow.
- [NEW] added two new sample configuration files that use flow. One is a simple unbounded region with a single flow vector. The second is a hybrid box with a different flow vector in each region.
- [ENHANCEMENT] increased MATLAB import time by about an order of magnitude. Simulation output files are now fully scanned in a single call, instead of being read line-by-line.
- [ENHANCEMENT] simplified microscopic algorithm to detect whether a given type of molecule is able to move in a given region
- [FIX] improved algorithm that places mesoscopic molecules in the microscopic regime so that the proposed molecule trajectory is checked for boundary collisions before the molecule is placed.
- [FIX] added import of local (region) diffusion coefficients to MATLAB
- [FIX] corrected the default subvolume base size if it is defined incorrectly
- [FIX] corrected the detection of a missing surface reaction probability for surface reactions

Changes introduced in AcCoRD v1.0 (2016-10-31)

src/actor.h
/*
 * The AcCoRD Simulator
 * (Actor-based Communication via Reaction-Diffusion)
 *
 * Copyright 2016 Adam Noel. All rights reserved.
 * 
 * For license details, read LICENSE.txt in the root AcCoRD directory
 * For user documentation, read README.txt in the root AcCoRD directory
 *
 * actor.h - operations on array of actors and its elements
 *
 * Last revised for AcCoRD v1.0 (2016-10-31)
 *
 * Revision history:
 *
 * Revision v1.0 (2016-10-31)
 * - added BURST modulation, which does not modulate binary data but always releases
 * molecules (of all types specified)
 * - moved mesoscopic structure fields from subvolume struct to meso subvolume struct
 *
 * Revision v0.6 (public beta, 2016-05-30)
 * - modified random number generation. Now use PCG via a separate interface file.
 * - added active point sources. Can be placed in microscopic or mesoscopic regions.
 * Cannot be on boundary of 2 or more regions or mesoscopic subvolumes
 * - added second order chemical reactions in the microscopic regime via a binding radius
 * and unbinding radius (can also model molecular crowding)
 * - made output of active actor data sequence a user option
 * - added bBits array for user to define a constant active actor bit sequence
 *
 * Revision v0.5.1 (2016-05-06)
 * - updated call to bPointInRegionNotChild to not exclude surface regions
 *
 * Revision v0.5 (2016-04-15)
 * - added ability to define location of actor by a list of regions
 * - added 2D and surface regions. Regions that have an effective dimension different
 * from their actual dimension cannot be intersected by a actor boundary (such regions
 * must be fully inside). Molecules will not be placed by an actor on a 2D region if
 * the actor overlaps at least 1 3D region
 * - tidied up calculations of subvolume coordinates
 *
 * Revision v0.4.1
 * - improved use and format of error messages
 *
 * Revision v0.4
 * - modified use of unsigned long and unsigned long long to uint32_t and uint64_t
 * - removed deprecated debug function
 * - added accommodation of spherical (microscopic) subvolumes and actors
 * - renamed region structure and array of structures to accommodate naming
 * of new boundary structure
 *
 * Revision v0.3.1
 * - header added
*/

#ifndef ACTOR_H
#define ACTOR_H

#include <stdio.h>
#include <stdlib.h> // for exit(), malloc
#include <stdbool.h> // for C++ bool naming, requires C99
#include <limits.h> // For SHRT_MAX
#include <string.h> // for strlen(), strcmp()
#include "region.h"
#include "base.h" // For region adjacency
#include "micro_molecule.h" // For molecule coordinate linked list
#include "global_param.h" // For region adjacency
#include "mol_release.h" // For each active actor's linked list of active emissions
#include "actor_data.h" // For each active actor's linked list of binary data
#include "subvolume.h"  // For subvolume structure

/*
* Data Type Declarations
*/

/* The actorStructSpec3D structure contains the user-defined actor parameters
* for a single 3D actor
*/
struct actorStructSpec3D { // Configuration parameters
	
	/*
	*  Parameters Common to Any Actor
	*/
	
	// Is the actor's location defined by a list of regions?
	// If true, then there must be a list of regions that define the location of the actor
	// If false, then the actor needs a shape and boundary specified
	bool bDefinedByRegions;
	
	// Number of regions that define the actor's location
	// Used if bDefinedByRegions == true
	short numRegion;
	
	// Names of regions that define the actor's location
	// Length is numRegionDefined
	char ** regionLabel;
	
	// Actor shape type
	// Values defined in global_param.h
	// Used if bDefinedByRegions == false
	int shape;
	
	// Outer boundary of this actor, defined in absolute space
	// TODO: Account for more actor shapes
	// Used if bDefinedByRegions == false
	double boundary[6];
	
	// Actors are "Active" or "Passive"
	// Active -> change a state of the simulation environment
	// Passive -> observe a state of the simulation environment
	bool bActive;
	
	// Actor's "Start time", i.e., when it is first supposed to act
	double startTime;
	
	// Does actor have a maximum number of times to act?
	bool bMaxAction;
	
	// Maximum number of actions (if bMaxAction == true)
	uint32_t numMaxAction;
	
	// Is actor behavior independent of the system state?
	bool bIndependent;
	
	// Frequency of independent behavior (if bIndependent == true)
	double actionInterval;
	
	// Record observations of actor to output file
	bool bWrite;
	
	/*
	*  Parameters for Active Actors
	*/
	
	// Is the number of molecules released stochastic?
	bool bNumReleaseRand;
	
	// Are the molecule release times stochastic?
	bool bTimeReleaseRand;
	
	// Period for release activity associated with one actionInterval
	// Could (in general) be less than or greater than actionInterval
	double releaseInterval;
	
	// Slot period (for "slotted" release, as opposed to kinetic)
	double slotInterval;
	
	// Is the actor activity defined by independent random bits?
	bool bRandBits;
	
	// Array of pre-defined bits
	// Defined if bRandBits == FALSE
	bool * bBits;
	
	// If bRandBits, what is the probability of one bit having value 1
	double probOne;
	
	// Modulation scheme. Values are defined in global_param.h
	unsigned short modScheme;
	
	// Modulation bits (i.e., modulation order is 2^modBits)
	unsigned short modBits;
	
	// Modulation base strength. Exact meaning depends on the
	// modulation scheme
	double modStrength;
	
	// Which types of molecules are released?
	// The number of types with value 1 should be consistent with the modulation scheme
	bool * bReleaseMol;
	
	// TODO: Consider parameters for recording the actions of this actor (i.e., the
	// randomly-modulated bits, random signal strength, action times, etc)
		
	/*
	*  Parameters for Passive Actors
	*/
	
	// Log observation time (if bWrite == true)
	bool bRecordTime;
	
	// Which molecule types are observed? (if bWrite == true)
	bool * bRecordMol;
	
	// Which molecule types have positions recorded?
	// (if bWrite == true AND bRecordMol[ID] == true)
	bool * bRecordPos;
};

/* The actorStruct3D structure contains all parameters specific to any 3D
* single simulation actor.
*/
struct actorStruct3D { // Common actor parameters

	// The user-defined common actor parameters
	struct actorStructSpec3D spec;

	//
	// "Initialization" parameters (Determined from configuration parameters at runtime)
	//
	
	// IDs of actor in active/passive list
	short activeID;
	short passiveID;
	
	// Size of actor
	// TODO: May not need to be a permanent structure member; initialization only
	double volume;
	
	// The number of regions that are (at least partially) within actor space
	unsigned short numRegion;
	
	// What is the largest effective dimension of any region within actor space?
	// If 3, then only normal 3D regions count towards actor volume
	// If 2, then only 3D surface regions and normal 2D regions count towards actor volume
	unsigned short maxDim;
	
	// The number of regions are (at least partially) within actor space AND
	// have the same effective dimension as the actor itself
	unsigned short numRegionDim;
	
	// Array of IDs of regions that are (at least partially) within actor space
	// Length is numRegion
	unsigned short * regionID;
	
	// Array of bools indicating whether the actor is fully inside an
	// overlapping region or not
	// Only applies if the actor is spherical, and if false then the region is
	// fully inside the actor
	// Length is numRegion
	bool * bRegionInside;
	
	// The number of subvolumes in each region that are (at least partially)
	// within actor space. Only applies if region is mesoscopic.
	// Length is numRegion
	uint32_t * numSub;
	
	// Shape of intersection of actor and region
	// Length is numRegion
	unsigned short * regionInterType;
	
	// Outer boundary of intersection of actor and region
	// TODO: May not need to be a permanent COMMON structure member; initialization
	// only (unless region is MICRO)
	// Length is numRegion x 6
	double (* regionInterBound)[6];
	
	// Area of intersection of actor and region
	// Length is numRegion
	// TODO: May not need to be a permanent structure member; initialization only
	double * regionInterArea;
	
	// Fraction of actor in intersection of actor and region
	// Length is numRegion
	double * cumFracActorInRegion;
	
	// The IDs of subvolumes in each region that are (at least partially)
	// within actor space. Only applies if region is mesoscopic.
	// Length is numRegion x numSub
	uint32_t ** subID;
	
	//
	// "Simulation" parameters (Determined at simulation time)
	//
	
	// NOTE: These parameters would need to be reset for each repeat

	// (Current) next "Action" time of actor
	double nextTime;
	
	// Index of current action (needed if bMaxAction == true)
	uint32_t curAction;
	
	// FUTURE MEMBERS (POTENTIAL)
	// Indicator for how next time is determined
	// Controls for dependent actors
};

/* The actorActiveStruct3D structure contains all parameters specific to an active 3D
* simulation actor.
*/
struct actorActiveStruct3D { // Active actor parameters

	//
	// "Initialization" parameters (Determined from configuration parameters at runtime)
	//
	
	// ID of actor in common (active and passive) list
	short actorID;
	
	// Size of symbol alphabet
	unsigned int alphabetSize;
	
	// Number of types of molecules that actor could release
	unsigned short numMolType;
	
	// Indices of the molecule types that actor could release
	// NOTE: The molecule types need to be defined in the order corresponding to the bit
	unsigned short * molType;
	
	// The cumulative fraction of actor that is within a given mesoscopic subvolume, given that the
	// subvolume is within actor space. Only applies if region is mesoscopic.
	// Length is numRegion x numSub
	double ** cumFracActorInSub;
	
	//
	// "Simulation" parameters (Determined at simulation time)
	//
	
	// Current actor bit when sequence is defined by user (i.e., bRandBits == false)
	uint32_t curBit;
	
	// Time for start of next NEW release by the actor (i.e., start of next symbol interval)
	double nextNewReleaseTime;
	
	// Time for next emission by a current release of the actor
	double nextEmissionTime;
	
	// Is next action time for a new symbol?
	bool bNextActionNewRelease;

	// Index of next emission by a current release of the actor
	unsigned int nextEmissionIndex;
	
	// Linked list for currently active release intervals
	ListRelease releaseList;
	
	// Random data associated with the actor
	ListData binaryData;
};

/* The actorPassiveStruct3D structure contains all parameters specific to a passive 3D
* simulation actor.
*/
struct actorPassiveStruct3D { // Passive actor parameters

	//
	// "Initialization" parameters (Determined from configuration parameters at runtime)
	//
	
	// ID of actor in common (active and passive) list
	short actorID;
	
	// The fraction of subvolume that is within actor space, given that a mesoscopic
	// subvolume is within actor space. Only applies if region is mesoscopic.
	// Length is numRegion x numSub
	double ** fracSubInActor;
	
	// Index of actor in list of passive actors that record observations
	short recordID;
	
	// Are we writing the position of at LEAST one molecule type in a mesoscopic region?
	// Indicates whether we need to define subInterBound
	// Length is numRegion
	bool * bRecordMesoAnyPos;
	
	// Outer boundary of intersection of actor and subvolume
	// Only allocated if region is mesoscopic and molecule positions are being recorded
	// Length is numRegion x numSub x 6
	double (** subInterBound)[6];
	
	// Number of molecule types that this actor records (if bWrite == true)
	unsigned short numMolRecordID;
	
	// IDs of molecules that this actor records counts of (if bWrite == true)
	// Length is numMolRecordID
	unsigned short * molRecordID;
	
	// Number of molecule types that this actor records positions
	// (if bWrite == true AND bRecordMol == true)
	unsigned short numMolRecordPosID;
	
	// IDs of molecules that this actor records positions of
	// (if bWrite == true AND bRecordMol == true)
	// Length is numMolRecordPosID
	unsigned short * molRecordPosID;
	
	//
	// "Simulation" parameters (Determined at simulation time)
	//
	
	// Number of molecules observed in the current observation
	// Length is numMolRecordID
	uint64_t * curMolObs;
};

/*
* Function Declarations
*/

void allocateActorCommonArray(const short NUM_ACTORS,
	struct actorStruct3D ** actorCommonArray);

void initializeActorCommon(const short NUM_ACTORS,
	struct actorStruct3D actorCommonArray[],
	const struct actorStructSpec3D actorCommonSpecArray[],
	const struct region regionArray[],
	const short NUM_REGIONS,
	short * NUM_ACTORS_ACTIVE,
	short * numActiveRecord,
	short ** activeRecordID,
	short * NUM_ACTORS_PASSIVE,
	short * numPassiveRecord,
	short ** passiveRecordID,
	const struct subvolume3D subvolArray[],
	uint32_t **** subID,
	uint32_t subCoorInd[][3],
	const double SUBVOL_BASE_SIZE);

void allocateActorActivePassiveArray(const short NUM_ACTORS_ACTIVE,
	struct actorActiveStruct3D ** actorActiveArray,
	const short NUM_ACTORS_PASSIVE,
	struct actorPassiveStruct3D ** actorPassiveArray);

void initializeActorActivePassive(const short NUM_ACTORS,
	const struct actorStruct3D actorCommonArray[],
	const unsigned short NUM_MOL_TYPES,
	const struct region regionArray[],
	const short NUM_REGIONS,
	const struct mesoSubvolume3D mesoSubArray[],
	const short NUM_ACTORS_ACTIVE,
	struct actorActiveStruct3D actorActiveArray[],
	const short NUM_ACTORS_PASSIVE,
	struct actorPassiveStruct3D actorPassiveArray[],
	uint32_t subCoorInd[][3]);

void resetActors(const short NUM_ACTORS,
	struct actorStruct3D actorCommonArray[],
	const unsigned short NUM_MOL_TYPES,
	const struct region regionArray[],
	const short NUM_REGIONS,
	const short NUM_ACTORS_ACTIVE,
	struct actorActiveStruct3D actorActiveArray[],
	const short NUM_ACTORS_PASSIVE,
	struct actorPassiveStruct3D actorPassiveArray[]);

void deleteActor(const short NUM_ACTORS,
	struct actorStruct3D actorCommonArray[],
	const struct region regionArray[],
	const short NUM_ACTORS_ACTIVE,
	struct actorActiveStruct3D actorActiveArray[],
	const short NUM_ACTORS_PASSIVE,
	struct actorPassiveStruct3D actorPassiveArray[],
	short passiveRecordID[],
	short activeRecordID[]);

// TODO: Move these release functions into a "higher" level file
	
void newRelease(const struct actorStruct3D * actorCommon,
	struct actorActiveStruct3D * actorActive,
	double curTime);

void findNextEmission(const struct actorStruct3D * actorCommon,
	struct actorActiveStruct3D * actorActive);

void fireEmission(const struct actorStruct3D * actorCommon,
	struct actorActiveStruct3D * actorActive,
	struct region regionArray[],
	const short NUM_REGIONS,
	struct subvolume3D subvolArray[],
	struct mesoSubvolume3D * mesoSubArray,
	const uint32_t numMesoSub,
	const unsigned short NUM_MOL_TYPES,
	ListMolRecent3D microMolListRecent[NUM_REGIONS][NUM_MOL_TYPES],
	double tMicro,
	uint32_t * heap_subvolID,
	uint32_t (*heap_childID)[2],
	bool (*b_heap_childValid)[2]);
	
void placeMolecules(const struct actorStruct3D * actorCommon,
	const struct actorActiveStruct3D * actorActive,
	struct region regionArray[],
	const short NUM_REGIONS,
	struct subvolume3D subvolArray[],
	struct mesoSubvolume3D * mesoSubArray,
	const uint32_t numMesoSub,
	uint64_t numNewMol,
	const unsigned short curMolType,
	const unsigned short NUM_MOL_TYPES,
	ListMolRecent3D microMolListRecent[NUM_REGIONS][NUM_MOL_TYPES],
	double tCur,
	double tMicro,
	uint32_t * heap_subvolID,
	uint32_t (*heap_childID)[2],
	bool (*b_heap_childValid)[2]);

void placeMoleculesInRegion(const struct actorStruct3D * actorCommon,
	const struct actorActiveStruct3D * actorActive,
	struct region regionArray[],
	const short curRegion,
	const short curRegionInter,
	const short NUM_REGIONS,
	struct subvolume3D subvolArray[],
	struct mesoSubvolume3D * mesoSubArray,
	const uint32_t numMesoSub,
	uint64_t numNewMol,
	const unsigned short curMolType,
	const unsigned short NUM_MOL_TYPES,
	ListMolRecent3D * microMolListRecent,
	double tCur,
	double tMicro,
	uint32_t * heap_subvolID,
	uint32_t (*heap_childID)[2],
	bool (*b_heap_childValid)[2]);
	
void placeMoleculesInSub(struct region regionArray[],
	struct subvolume3D subvolArray[],
	struct mesoSubvolume3D * mesoSubArray,
	const uint32_t numMesoSub,
	uint64_t numNewMol,
	const unsigned short curMolType,
	const uint32_t curMeso,
	const unsigned short NUM_MOL_TYPES,
	double tCur,
	const short NUM_REGIONS,
	uint32_t * heap_subvolID,
	uint32_t (*heap_childID)[2],
	bool (*b_heap_childValid)[2]);

// Find range of subvolumes to search over for intersection with an actor
void findSubSearchRange(const struct region regionArray[],
	const short curRegion,
	const short curInterRegion,
	const struct actorStruct3D actorCommonArray[],
	const short curActor,
	uint32_t * first1,
	uint32_t * first2,
	uint32_t * first3,
	uint32_t * last1,
	uint32_t * last2,
	uint32_t * last3,
	bool bHaveCur1,
	uint32_t cur1);

#endif // ACTOR_H
